cpp/src/quantiles/quantiles.cu

```diff
@@ -1,109 +1,90 @@
-            auto result = get_array_value<ScalarResult>(input.begin<T>(), 0);
-            return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
-        }
+    auto quantile_idx_lookup = [sortmap_lookup, interp, size=input.num_rows()]
+        __device__ (double q) {
+            return detail::select_quantile<size_type>(sortmap_lookup,
+                                                      size,
+                                                      q,
+                                                      interp);
+        };
 
-        auto valid_count = input.size() - input.null_count();
+    rmm::device_vector<double> q_device = q;
 
-        if (not column_order.is_ordered)
-        {
-            table_view const in_table { { input } };
-            auto sortmap = sorted_order(in_table, { order::ASCENDING }, { null_order::AFTER });
-            auto sortmap_begin = sortmap->view().begin<size_type>();
-            auto input_begin = input.begin<T>();
+    auto quantile_idx_iter = thrust::make_transform_iterator(q_device.begin(),
+                                                             quantile_idx_lookup);
 
-            auto selector = [&](size_type location) {
-                auto idx = detail::get_array_value<size_type>(sortmap_begin, location);
-                return detail::get_array_value<T>(input_begin, idx);
-            };
+    return detail::gather(input,
+                          quantile_idx_iter,
+                          quantile_idx_iter + q.size(),
+                          false,
+                          false,
+                          false,
+                          mr);
+}
 
-            auto result = select_quantile<ScalarResult>(selector, valid_count, percent, interp);
-            return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
-        }
+}
 
-        auto input_begin = column_order.ordering == order::ASCENDING
-            ? input.begin<T>() + (column_order.null_ordering == null_order::BEFORE ? input.null_count() : 0)
-            : input.begin<T>() - (column_order.null_ordering == null_order::AFTER  ? input.null_count() : 0) + input.size() - 1;
+std::unique_ptr<table>
+quantiles(table_view const& input,
+          std::vector<double> const& q,
+          interpolation interp,
+          cudf::sorted is_input_sorted,
+          std::vector<order> const& column_order,
+          std::vector<null_order> const& null_precedence,
+          rmm::mr::device_memory_resource* mr)
+{
+    CUDF_EXPECTS(interp == interpolation::HIGHER ||
+                 interp == interpolation::LOWER ||
+                 interp == interpolation::NEAREST,
+                 "multi-column quantiles require a non-arithmetic interpolation strategy.");
 
-        auto selector = [&](size_type location) {
-            return get_array_value<T>(input_begin, column_order.ordering == order::ASCENDING ? location : -location);
-        };
+    CUDF_EXPECTS(input.num_rows() > 0,
```

Comment:
Isn't this a case where we should return an empty table rather than throwing? Dask may produce an empty partition, right?

Reply:
Jake and I were discussing it. I should have kept that comment open a while longer. Returning `empty_like` sounds good to me. @jrhemstad Do you feel strongly either way?

Reply:
I thought we discussed that if the quantile vector is empty that should be an exception. Empty inputs should always return empty outputs.

Reply:
Wait, this is not what Jake and I discussed. Sorry.

This is an error case, IMO. Quantiles are undefined for empty inputs, and returning an empty table does not provide any new information to the consumer. Consumers should expect to receive the same number of rows as the number of quantiles they request.

Reply:
I'm fine with updating this to `return empty_like(input)` when `num_rows() == 0`, if that's what's required from an ETL integration perspective, but from a purely mathematical standpoint, asking for the quantiles of an empty collection is an undefined operation.

Reply:
In contrast, asking for zero quantiles of a non-empty collection is a well-defined operation, so I feel the advice here is backwards, unless it's required for integration purposes.

Reply:
For instance, if we assert `q.size() > 0`, we are preventing the user from performing a well-defined operation, but if we _don't_ assert that `input.num_rows() > 0`, we are allowing the user to perform an undefined operation, but making up for it by returning a fake result.

Reply:
Dask frequently provides empty inputs in order to determine the type of the outputs. CC @kkraus14 

> In contrast, asking for zero quantiles of a non-empty collection is a well-defined operation (return empty), 

Sure, this makes sense to me. If `q.empty()` just return an empty table.

Reply:
Does Dask interact directly with libcudf? Might this sort of massaging be better suited to the python layer? The dividing line between implementation correctness and etl friendliness is a bit fuzzy for me.

Reply:
> but from a purely mathematical standpoint, asking for the quantiles of an empty collection is an undefined operation.

Why? If I ask for the median of an empty set, it makes sense to me that it returns an empty set.

Reply:
If a quantile represents a divide in a collection, but the collection is indivisible, it cannot exist and is therefore undefined.

Reply:
Yeah, makes sense. I'll leave the decision up to the Python layer folks

Reply:
@jrhemstad Are we leaving the cpp impl up to python folks, or suggesting the cython layer should adapt?

Reply:
@kkraus14 please comment.

Reply:
Pandas throws here so I'm fine with whatever you want to do on the libcudf side. We'll likely short circuit calling the API if they ask for the quantile of an empty table regardless.